test/logflare/backends/adaptor/datadog_adaptor_test.exs

```diff
@@ -0,0 +1,117 @@
+defmodule Logflare.Backends.Adaptor.DatadogAdaptorTest do
+  use Logflare.DataCase, async: false
+
+  alias Logflare.Backends.Adaptor
+
+  @subject Logflare.Backends.Adaptor.DatadogAdaptor
+
+  doctest @subject
+
+  describe "cast and validate" do
+    test "API key is required" do
+      changeset = Adaptor.cast_and_validate_config(@subject, %{})
+
+      refute changeset.valid?
+
+      changeset =
+        Adaptor.cast_and_validate_config(@subject, %{
+          "api_key" => "foobarbaz"
+        })
+
+      assert changeset.valid?
+    end
+  end
+
+  describe "ingest" do
+    setup do
+      this = self()
+
+      TestServer.add("/",
+        via: :post,
+        match: fn conn ->
+          {:ok, body, conn} = Plug.Conn.read_body(conn)
+          events = Jason.decode(body)
+
+          send(this, events)
+
+          Plug.Conn.send_resp(conn, 200, "")
+        end
+      )
+
+      pid =
+        start_supervised!({
+          @subject.Client,
+          timeout: 100, threshold: 3, url: TestServer.url(), headers: [{"foo", "bar"}]
+        })
+
+      {:ok, pid: pid}
+    end
+
+    test "after timeout it sends messages", %{pid: pid} do
+      events = [
+        %Logflare.LogEvent{
+          body: %{event_message: "hello world"},
+          source: %{name: "sample"}
+        }
+      ]
+
+      @subject.ingest(pid, events, [])
+
+      assert_receive {:ok, resp}, 200
+      assert length(resp) == 1
+    end
+
+    test "after sent if there is more events than threshold", %{pid: pid} do
+      events =
+        List.duplicate(
+          %Logflare.LogEvent{
+            body: %{event_message: "hello world"},
+            source: %{name: "sample"}
+          },
+          5
+        )
+
+      @subject.ingest(pid, events, [])
+
+      assert_receive {:ok, resp}, 200
+      assert length(resp) == 5
+    end
+
+    test "after sent if there is more events than threshold after second ingestion", %{pid: pid} do
+      events =
+        List.duplicate(
+          %Logflare.LogEvent{
+            body: %{event_message: "hello world"},
+            source: %{name: "sample"}
+          },
+          2
+        )
+
+      @subject.ingest(pid, events, [])
+      @subject.ingest(pid, events, [])
+
+      assert_receive {:ok, resp}, 200
+      assert length(resp) == 4
+    end
+
+    test "metadata other than `event_message` is stored in `ddtags`", %{pid: pid} do
+      events = [
+        %Logflare.LogEvent{
+          body: %{
+            event_message: "hello world",
+            foo: "foo",
+            bar: %{baz: 4}
+          },
+          source: %{name: "sample"}
+        }
+      ]
+
+      @subject.ingest(pid, events, [])
+
+      assert_receive {:ok, [event]}, 200
+      assert %{"ddtags" => tags} = event
+      assert tags =~ ~r/foo:foo/
```

Comment:
tags need to be a certain format
https://docs.datadoghq.com/getting_started/tagging/#define-tags
would need to handle restrictions for them.

Reply:
Ok, I will take a look. Maybe there should a way to define mappings for tags to avoid excessive tag amount as docs mention:

> Tags should not originate from unbounded sources, such as epoch timestamps, user IDs, or request IDs. Doing so may infinitely [increase the number of metrics](https://docs.datadoghq.com/metrics/) for your organization and impact your billing.

Reply:
yea i'm thinking of implementing a allowlist for schema paths that are permitted for schema updates, it can take a similar form:
```
metadata\..+
my
field
```
was thinking of making it regex based, can do a similar allowlist for allowed tag mappings